### PR TITLE
SPOI-8090 #resolve #comment Add Dimension Computation into benchmark

### PR DESCRIPTION
**Pull request by brightchen**

### CODE REVIEW

**Comment by brightchen** [comment]
@sandeshh please review and merge

**Comment by brightchen** [comment]
@sandeshh please don't merge. This PR should be merged after Megh 3.4 released.

File: apex-benchmarks/src/main/java/apex/benchmark/DimensionTupleGenerator.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+ */
+package apex.benchmark;
+
+import java.util.Random;
+
+public class DimensionTupleGenerator
+{
+  public static final String[] adIds = {"ad1", "ad2", "ad3", "ad4", "ad5", "ad6", "ad7", "ad8", "ad9", "ad10" };
+  public static final String[] campaignIds = {"campaign1", "campaign2", "campaign3", "campaign4", "campaign5", "campaign6", "campaign7", "campaign8", "campaign9", "campaign10" };
+  public static final Long[] eventTimeShifts = {-25000L, -15000L, 0L};
```
Comment (sandeshh) [inline apex-benchmarks/src/main/java/apex/benchmark/DimensionTupleGenerator.java]: Number of Campaigns and Ads should be configurable.

**Comment by sandeshh** [comment]
Keep only the application that will be used for benchmarking

**Comment by brightchen** [comment]
@sandeshh please review and merge

File: apex-benchmarks/src/main/java/apex/benchmark/JasonEventGenerator.java [hunk 2]
```diff
@@ -0,0 +1,125 @@
+/**
```
Comment (sandeshh) [inline apex-benchmarks/src/main/java/apex/benchmark/JasonEventGenerator.java]: Why do you need JasonEventGenerator, when there is JasonGenerator ?

File: apex-benchmarks/src/main/java/apex/benchmark/ConfigUtil.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+  
+  public static String getGatewayAddress(DAG dag, Configuration conf)
+  {
+    String gatewayAddress = dag.getValue(DAGContext.GATEWAY_CONNECT_ADDRESS);
+    if (gatewayAddress == null) {
+      gatewayAddress = conf.get(PROP_GATEWAY_ADDRESS);
+    }
+    return gatewayAddress;
+  }
+}
```
Comment (sandeshh) [inline apex-benchmarks/src/main/java/apex/benchmark/ConfigUtil.java:35]: new line required for all the files.

File: apex-benchmarks/src/main/java/apex/benchmark/ApplicationDimensionComputation.java [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+public class ApplicationDimensionComputation implements StreamingApplication
+{
+  public static final String APP_NAME = "DimensionComputation";
+  public static final String DIMENSION_SCHEMA = "eventSchema.json";
+  private static final transient Logger logger = LoggerFactory.getLogger(ApplicationDimensionComputation.class);
+  
+  protected static final int PARTITION_NUM = 8;
+  
+  protected String appName;
+
```
Comment (sandeshh) [inline apex-benchmarks/src/main/java/apex/benchmark/ApplicationDimensionComputation.java]: Variable not used?

File: apex-benchmarks/src/main/java/apex/benchmark/ApplicationDimensionComputation.java [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+@ApplicationAnnotation(name = ApplicationDimensionComputation.APP_NAME)
+public class ApplicationDimensionComputation implements StreamingApplication
+{
+  public static final String APP_NAME = "DimensionComputation";
+  public static final String DIMENSION_SCHEMA = "eventSchema.json";
+  private static final transient Logger logger = LoggerFactory.getLogger(ApplicationDimensionComputation.class);
+  
+  protected static final int PARTITION_NUM = 8;
+  
+  protected String appName;
```
Comment (sandeshh) [inline apex-benchmarks/src/main/java/apex/benchmark/ApplicationDimensionComputation.java]: appName variable not used

File: apex-benchmarks/src/main/java/apex/benchmark/ApplicationDimensionComputation.java [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+  }
+
+  public void populateDimensionsDAG(DAG dag, Configuration conf, DefaultOutputPort<DimensionTuple> upstreamPort) 
+  {
+    final String eventSchema = SchemaUtils.jarResourceFileToString(eventSchemaLocation);
+    
+    // dimension
+    DimensionsComputationFlexibleSingleSchemaPOJO dimensions = dag.addOperator("DimensionsComputation",
+        DimensionsComputationFlexibleSingleSchemaPOJO.class);
+    dag.getMeta(dimensions).getAttributes().put(Context.OperatorContext.APPLICATION_WINDOW_COUNT, 10);
```
Comment (sandeshh) [inline apex-benchmarks/src/main/java/apex/benchmark/ApplicationDimensionComputation.java]: Why not use Dag.setAttributes everywhere ?

File: apex-benchmarks/src/main/java/apex/benchmark/ApplicationWithDCWithoutDeserializer.java [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+
+    dag.setAttribute(eventGenerator, Context.OperatorContext.PARTITIONER, new StatelessPartitioner<EventGenerator>(PARTITION_NUM));
+
+    return converter.outputPort;
+  }
+  
+  private void setupRedis(Map<String, List<String>> campaigns) {
+
+    RedisHelper redisHelper = new RedisHelper();
+    redisHelper.init("node35.morado.com");
```
Comment (sandeshh) [inline apex-benchmarks/src/main/java/apex/benchmark/ApplicationWithDCWithoutDeserializer.java]: remove the hardcoded location

File: apex-benchmarks/src/main/java/apex/benchmark/ConfigUtil.java (hunk 3 above)
Comment (brightchen) [inline apex-benchmarks/src/main/java/apex/benchmark/ConfigUtil.java:35]: it in fact ended with new line. you can checkout the code

**Comment by brightchen** [comment]
@sandeshh please review

**Pull request merged**